sdk/storage/Azure.Storage.DataMovement/src/Shared/TransferManager.cs

```diff
@@ -292,6 +294,63 @@ public virtual async Task PauseTransferIfRunningAsync(string transferId, Cancell
             }
         }
 
+        /// <summary>
+        /// Lists all the transfers stored in the checkpointer that can be resumed.
+        /// </summary>
+        /// <returns>
+        /// List of <see cref="DataTransferProperties"/> objects that can be used to rebuild resources
+        /// to resume with.
+        /// </returns>
+        public virtual async IAsyncEnumerable<DataTransferProperties> GetResumableTransfersAsync()
```

Comment:
Yes, each call to `GetByteValue` and `GetResourceIdAsync` is opening **the first** job part file once and each call to `GetResourcePathAsync` is opening every job part file in the transfer once. `GetStoredTransfersAsync` and `CurrentJobPartCountAsync` don't needs to open anything. So, for each transfer we are performing 3 + 2n (with n being the number of job parts) file accesses.

While I do agree this is pretty inefficient and I did think about this, changing this to a single read would add quite a bit of complexity with index calculations and all and we'd need to read a significant portion of the file. I could do it if you think it's worth it, but I do think a lot of this will change with the next schema revision where we have a register file of basic info on all the transfers. Ideally, we put paths and resource ids (and probably job status) in there.

There is a probably a happy medium here where I can combine the source and destination reads for resource id and paths to make the number of file accesses 2 + n instead. I will try that at least.

Reply:
I ended up doing the last thing I mentioned, reducing the file access count to 2 + n for each transfer. I think further optimization is not worth the complexity at this point (with this likely being replaced).